stdlib/universe/window2.go

```diff
@@ -0,0 +1,490 @@
+	indices = arrow.IntSlice(indices, start, stop)
+	defer indices.Release()
+
+	// Copy the values from the column reader.
+	for j, col := range b.Columns {
+		builder := b.Builders[j]
+
+		switch col.Label {
+		case w.startCol:
+			b := builder.(*array.Int64Builder)
+			b.Reserve(indices.Len())
+			for i, n := 0, indices.Len(); i < n; i++ {
+				b.Append(int64(bound.Start))
+			}
+		case w.stopCol:
+			b := builder.(*array.Int64Builder)
+			b.Reserve(indices.Len())
+			for i, n := 0, indices.Len(); i < n; i++ {
+				b.Append(int64(bound.Stop))
+			}
+		default:
+			idx := execute.ColIdx(col.Label, cr.Cols())
+			arr := table.Values(cr, idx)
+			arrowutil.CopyByIndexTo(builder, arr, indices)
+		}
+	}
+	return start
+}
+
+// getWindowSpan retrieves the span of indexes that fit into this boundary.
+// There will always be at least one point because we only invoke this method
+// with boundaries that we have already determined to have at least one row.
+func (w *windowTransformation2) getWindowSpan(ts, indexes *array.Int64, bound execute.Bounds, offset int) (start, stop int) {
+	// Find the start offset that fits in this boundary.
+	n := indexes.Len()
+	for ; offset < n; offset++ {
+		t := ts.Value(int(indexes.Value(offset)))
+		if values.Time(t) >= bound.Start {
+			break
+		}
+	}
+
+	// Determine the stop offset.
+	stop = offset + 1
+	for ; stop < n; stop++ {
+		t := ts.Value(int(indexes.Value(stop)))
+		if !bound.Contains(values.Time(t)) {
+			break
+		}
+	}
+	return offset, stop
+}
+
+func (w *windowTransformation2) Finish(id execute.DatasetID, err error) {
+	w.d.Finish(err)
+}
+
+func (w *windowTransformation2) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return w.d.RetractTable(key)
+}
+func (w *windowTransformation2) UpdateWatermark(id execute.DatasetID, t execute.Time) error {
+	return w.d.UpdateWatermark(t)
+}
+func (w *windowTransformation2) UpdateProcessingTime(id execute.DatasetID, t execute.Time) error {
+	return w.d.UpdateProcessingTime(t)
+}
+
+type windowSchemaTemplate struct {
+	keyCols   []flux.ColMeta
+	keyValues []values.Value
+	cols      []flux.ColMeta
+}
+
+// WindowSortedInputRule will insert a sort by time node in between
+// a group behavior-like node and a window node.
+//
+// This is done because the window algorithm is more efficient when it
+// is working on sorted data since it inserts and partitions data more
+// efficiently if points within the same window are close to each other
+// and not separated by too many buffers.
+//
+// This scenario happens often when group or keep/drop are used because
+// these transformations have a habit of regrouping the data in a way
+// that can cause the data to become unsorted.
+type WindowSortedInputRule struct{}
+
+func (w WindowSortedInputRule) Name() string {
+	return "WindowSortedInputRule"
+}
+
+func (w WindowSortedInputRule) Pattern() plan.Pattern {
+	return plan.Pat(WindowKind,
+		plan.OneOf([]plan.ProcedureKind{SchemaMutationKind, GroupKind}),
```

Comment:
This needs a `plan.Any()` as predecessor to the group/mutation.
```
return plan.Pat(
        WindowKind,
        plan.OneOf(
            []plan.ProcedureKind{SchemaMutationKind, GroupKind},
            plan.Any()))
```
Tried that out and there is a panic that occurs.
```
ts=2021-05-26T17:02:20.318730Z lvl=info msg="panic during program start" log_id=0UMPqR4G000 ot_trace_id=38359b56d810afcb ot_trace_sampled=false error="panic: Inconsistent plan graph: successor does not have edge back to predecessor" stacktrace="goroutine 182 [running]:\nruntime/debug.Stack
```

Reply:
Ah, ok. Do you know why the `plan.Any()` is needed?

Reply:
Fixed either way. When I tested against an influxdb instance with these optimizations enabled, I got this from the profiler:

```
$ influx query -o influxdata --profilers query,operator 'import "internal/debug" from(bucket: "test1") |> range(start: -2d) |> group() |> window(every: 1d)'
Result: _result
Table: keys: [_start, _stop]
                   _start:time                      _stop:time                      _time:time                  _value:float           _field:string     _measurement:string  
------------------------------  ------------------------------  ------------------------------  ----------------------------  ----------------------  ----------------------  
2021-05-26T00:00:00.000000000Z  2021-05-26T19:38:14.589442000Z  2021-05-26T18:58:09.275329000Z                             4                      f0                      m0  
Result: _profiler
Table: keys: [_measurement]
   _measurement:string             Concurrency:int         CompileDuration:int           QueueDuration:int            PlanDuration:int         RequeueDuration:int         ExecuteDuration:int           TotalDuration:int            MaxAllocated:int          TotalAllocated:int    RuntimeErrors:string                                                                                                                                                                       flux/query-plan:string  influxdb/scanned-bytes:int  influxdb/scanned-values:int  
----------------------  --------------------------  --------------------------  --------------------------  --------------------------  --------------------------  --------------------------  --------------------------  --------------------------  --------------------------  ----------------------  -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------  --------------------------  ---------------------------  
        profiler/query                           0                      188762                       16404                           0                           0                     6797899                     7021636                        2624                           0                          digraph {
  ReadGroup6
  // GroupMode: 2, GroupKeys: [], AggregateMethod: ""
  sort4
  window3
  generated_yield

  ReadGroup6 -> sort4
  sort4 -> window3
  window3 -> generated_yield
}

                           0                            0  
Table: keys: [_measurement]
   _measurement:string                      Type:string            Label:string                   Count:int             MinDuration:int             MaxDuration:int             DurationSum:int            MeanDuration:float  
----------------------  -------------------------------  ----------------------  --------------------------  --------------------------  --------------------------  --------------------------  ----------------------------  
     profiler/operator    *universe.sortTransformation2                   sort4                           1                       48018                       48018                       48018                         48018  
     profiler/operator        *influxdb.readGroupSource              ReadGroup6                           1                      236816                      236816                      236816                        236816  
     profiler/operator  *universe.windowTransformation2                 window3                           1                       41243                       41243                       41243                         41243  
```

Reply:
The empty predecessor pattern matches against nodes with  with no predecessors, so you have to put a pattern there.

Reply:
How does that compare to the before time in your setup? In IDPE I'm finding the additional sort between a pushed-down group and the window is slowing down total query time by a factor of 2.

Reply:
I am really feeling the feature flagging pain. Takes so long to get a before and after timing with a new feature disabled/enabled.

Reply:
I haven't tried the speed difference with a push down. Would it be better if I split out this planner rule? We can then separate this planner rule from the other performance bits since this planner rule wasn't required to get the speed ups we were saw.

Reply:
I've removed this now. I'd like to revisit it later but I don't think it's needed to make this worth merging.